### DISCUSSION TITLE
Project Continuation Steps

### DISCUSSION THREAD

**Discussion by mattsta** [mattsta/ib_insync#1] [Ideas]
So, what do we do now?

As far as I can tell, `ib_insync` was a single-owner project and now the owner is reportedly dead (who updated their github profile and closed all the projects though?).

We are locked out of the official documentation site and the pypi package and any existing testing infrastructure, so do we:

- rename this as a new project?
- keep the same project but bump a major version number to denote we're making our own changes now?
- should all this live in my own personal account? I have an unused github organization over here we could use too https://github.com/tradelogic
- we also need to audit the `ib_insync` API version and message compatibility against the current IBKR python API. There's supposedly more message types in newer IBKR gateways not supported in `ib_insync` yet. Ideally we should be watching for all changes to twsapi updates and making matching modifications to `ib_insync`
- Eventually it would be nice to have an extra _better_ client than just reproducing the IBKR api directly since it does awful things like use `FLOAT_MAX` and `INT_MAX` to mean `undefined` or `null` but we should be able to have a proper `None`-capable python API in more places.

Do we want to keep the same "trunk-branch development" process or use a more interactive `next` branch where we stage work until it gets tagged as a new version?

I'll start with some minor cleanup (common style formatter, pyproject instead of requirements and setup, etc), but we'll need a longer term plan for adding new features when IBKR APIs change plus continually monitoring for bug fixes along the way.

This project also relies on two of his other projects as utility libraries (`eventkit` and `nest_asyncio`) which also present the same problems: locked out of pypi and docs, etc, but I'll go ahead and start updating them with modern style refactoring just to get things moving.

**Comment by francdoc** [comment]
Hi @mattsta, you raised up some good issues here. You can share these questions in this discussion thread if you want: [[link to the thread](https://groups.io/g/twsapi/topic/104781778#52443)]. 

Some experienced developers are actively discussing similar issues (e.g: group organization, future architecture and design optimizations), and their insights could provide valuable guidance.

**Comment by salsasepp** [comment]
Thank you for your initiative! 
I guess you are aware of PEP 541, but anyway: https://peps.python.org/pep-0541/#how-to-request-a-name-transfer
PEP 541 specifies the process of taking over abandoned packages. The process it not trivial and certainly not fast. If successful, continuing the use of the ib_insync package name would be the least disruptive thing IMO. In the meantime, most of us can continue to use the latest pypi release, and you could distribute updated wheels outside of pypi if necessary.

**Comment by kulki007** [comment]
Hi @mattsta, Could you please respond on the thread linked above. It will  be better to come together as a community and decide how we should proceed further. Already ib_insyc is behind the latest IB API.

**Comment by mattsta** [comment]
The thread is amusing and sad at the same time. What part of `dead` and `[cross] 2024` are people misunderstanding?

Maybe he died and his estate has email/github access and shut things down. [Update: [they added a note and that's what happened](https://github.com/erdewit/ib_insync/commit/ef5ea29e44e0c40bbadbc16c2281b3ac58aa4a40)]

As for group participation, it's an open source curse (combined with the kinds of low knowledge/high maintenance people online trading attracts in general). If tens or hundreds or thousands of people demand software be maintained, but nobody is paying for it, you are at the largesse of generous experienced people to provide valuable work for free.

Feel free to have his sponsors bounce over to my account for more reliable updates. I plan on auditing the current feature difference between the IBKR `twsapi` package and `ib_insync` to see what's missing and also to see if we can maintain a cleaner automated update cycle when new TWS/gateway APIs are released.

**Comment by kulki007** [comment]
Mattsta, I am happy to help you in any way I can.is there anything specific that you need help with that I can provide. Immediate needs seems to be to move to latest 10.27 IB API.

**Comment by jlixfeld** [comment]
@mattsta - Are you following along in the [TWS API thread](https://groups.io/g/twsapi/topic/104781778#52443) by chance? I think at the moment there might be yourself and Mel who are trying to make changes to your respective forks, which is great but I wonder if some discussion or collab might be useful?  I think this might be a point where a potentially crucial divergence in the code based could stem from and maybe this doesn't have to be the case?

**Comment by kulki007** [comment]
@mattsta I agree with jixfeld. Instead of having two separate branches it would be better for the community at large to decide which branch to continue to work on. We are also trying to resurrect the ib_insync group on groups.io so that would be a place for people to discuss and ask questions about the API.

**Comment by mattsta** [comment]
Yeah, according to the `ibapi` version, the IBKR python API is using server version `180` while current `ib_insync` is on version `178`. The only additions appear to be "fund data fields" and "manual order time exercise options" (added October 2023) in the newer version IDs.

The "17 additional fields for FUND" is currently missing as people mention, but it's a specialized type of mutual fund metadata like this:

```python
class FundAssetType(Enum):                                                                                      
    NoneItem = ("None", "None")                                                                                 
    Others = (("000", "Others"),)                                                                               
    MoneyMarket = ("001", "Money Market")                                                                       
    FixedIncome = ("002", "Fixed Income")                                                                       
    MultiAsset = ("003", "Multi-asset")                                                                         
    Equity = ("004", "Equity")                                                                                  
    Sector = ("005", "Sector")                                                                                  
    Guaranteed = ("006", "Guaranteed")                                                                          
    Alternative = ("007", "Alternative")                                                                        
                                                                                                                
                                                                                                                
class FundDistributionPolicyIndicator(Enum):       
    NoneItem = ("None", "None")               
    AccumulationFund = ("N", "Accumulation Fund")
    IncomeFund = ("Y", "Income Fund") 
```

Should be easy enough to add. Maybe I'll have time over the next couple days to start clearing up these minor out of date features (or feel free to submit a PR for them too, adding these new `FUND` fields just requires repetitive template setup copying the logic from `tws-api/ibapi/contract.py` and `tws-api/ibapi/decoder.py` to fit the `ib_insync` data model instead).

Unfortunately due to IBKR's restrictive client API development practices, I can't just link to their usage in a github repo for us to use as a shared reference. I wonder if anybody cares if we create a new repo for the `twsapi` source too? Is there already one somewhere? They just make everybody click through an annoying EULA to download their zip files without any revision control: https://interactivebrokers.github.io/ — though they do have a "secret CLA repo" somewhere too https://interactivebrokers.github.io/api_software_contribute.html

As for overall continuity of the project, my primary concern is not breaking my personal [ibkr order and account management system](https://github.com/mattsta/icli) since it uses `ib_insync` as the IBKR API, so my priority is maintaining (and improving!) `ib_insync` code quality and features and performance so my trading interface and algo adapters don't decay over time. I'm less immediately concerned with packaging and doc updates since we are locked out of those anyway for who knows how long (still, assuming this entire "dead status + archive-all-the-things + delete forums" routine isn't just a hack, but is an indication of some final event or actively giving up on everything (or maybe he won a billion hague bucks and just doesn't have time for us anymore!), so we are on our own from now on).

Feel free to open any other topics/issues/PRs here and anybody can talk about anything. Probably a standard pattern of "Issues" for code questions and bug reports, "Discussions" for usage questions as long as we're not doing your homework for you, and "PRs" for, well, you know.

**Reply by ClimberMel** [reply]
Hi mattsa, it appears we have been working on similar paths... I invited people from groups.io to join my discord for additional discussions.  It has been around quite a while for collaborating on ib_insync projects.
We should decide on where to house the new fork and decide if it needs to be re-named to move forward.
I don't want to divide potential developers and their contributions.  It really doesn't matter where the code base is housed as long as it becomes a true community based open source software.  I'm of the opinion that what happened to the original author (while important to people in his life) is not as important as moving the project itself forward.  
Please join the discussions on discord as well: https://discord.gg/9w4hECKM

Are you willing / able to manage the github repo along with PRs and issues?
How familiar are you with transitioning the pypi source and changing over the readthedocs?
The other question is the licence.  It states that Ewald's licence must be kept with the code...
Thanks,
Mel

**Reply by mattsta** [reply]
Indeed, we have some priorities to work through:

1. continuity of code quality / features / maintenance
2. continuity of distribution (currently the last pypi release was almost a year ago and nobody has any large feature or bug complaints so far: https://libraries.io/pypi/ib-insync)
    - and the current package is downloaded around 1,000 times per day? how do we notify all those people? or, it could be just 1 user with an aggressive CI system and no local cache https://pypistats.org/packages/ib-insync — also according to the stats it would be safe to increase the minimum supported Python version from 3.6 currently to at least 3.8 because the most downloads are now from `(3.8, 3.11, 3.10, 3.9)`)
4. continuity of documentation site

I think changing the project name just depends on whether we want to try and claim the pypi package and the readthedocs site. If we can't get those, then we have to change the name anyway to publish anything new.

The current project name is perfect for its usage described as `An IB component that automatically keeps in sync with the TWS or IB Gateway application` but if people want to suggest potentially new names we blast into a new org/repo structure feel free to do so.

The problem with renaming is there's already a scattered userbase with 3 tiers of involved parties:

- code contributors (historically, just 1 primary owner/maintainer and random bug reports PRs from 3-5 other people per year)
- "interested" people like group/mailing list watchers
- hundreds or thousands of anonymous users who just download-and-go but never interact with the project

and we see people also requesting new governance structures and policies and other control-level things, but this has always been a low resource unfunded single-owner project so it's very much "you get what you pay for."

**Reply by kulki007** [reply]
I have been doing some research into reclaiming that project name. I believe its going to be a pain in the ass. It would be easier to simply change the name of the project. Over time people will figure out the new project. Transition will be a choppy but its a clean break. Hopefully we wont just have one owner but a few different owners who can maintain this code. 
I agree that the current project name is perfect and any other name is just not going to be that great. My suggestion was something like ib_wrapper. word IB needs to be in there to make it obvious that is related to IB.

**Reply by jlixfeld** [reply]
@mattsta I don't get the impression it's going to be trivial to claim the pypi and the rtd site.  I think renaming is the direction.  I think this is a critical point to reach consensus on in the spirit of being able to then be able to move forward with continuing development under one banner and doing the best that we can to collect as much of the scattered tiers of users as possible.

I recognize and appreciate that your interests are primarily in the stability of `icli`, but certainly want to recognize your skills, and insights and especially your contributions to the project; I would hate to see a decision made without giving you the opportunity to opt-in or out.  I of course have no control over what the ultimate outcome is, I'm just trying my best to rally as many people as I can in as diplomatic manner as I can based on what little I know and how little influence I have in the grand scheme of things :)

In the spirit of new names, I'll share some of my views and those oof some of the folks on Mel's Discord:

ib-api-reloaded
ib-insync-reloaded
ibapi_async
ib_wrapper
ib-tws-gw-api

The last one is my contribution.  My thoughts around that name stem from my personal feeling that it's important to articulate exactly what the library does, not necessary how it does it.  The name reflects that it's for IB, that it services the TWS and GW (as opposed to the Client Portal API) and that it itself it's an API library.  It intentionally lacks anything about sync or insync because for lack of any better understanding, that part of the name relates to its async functionality and I don't think it's overly important for a user to know, from the name, that it uses async.  Others have suggested that insync has more to do with its capabilities of synchronizing state with IB.  Again, like async, those are useful pieces of information to provide in a README.md, but I don't think it needs to be part of the name.

Finally, to one of your previous points, I think an umbrella GitHub organization is probably a good idea.  I feel like the organization name and slug should, in kind, reflect the projects that it encompasses.

**Reply by ReliefCrew** [reply]
I'm surprised you didn't suggest ib_async. Anyway, "what's in a name?"... apparently a lot: https://peps.python.org/pep-0423/#how-to-rename-a-project

**Reply by jlixfeld** [reply]
Indeed.  It would definitely be simpler to be able to just take on the old project, but that seems impossible to do in a reasonable amount of time and volunteer efforts.  Within the time it would take to make that so, I, naively, think that the community will likely be so fractured and disconnected from one another that it would likely do considerable harm.  Am I wrong?

**Reply by mattsta** [reply]
one of those "it was right in front of us the entire time!" moments — I think `ib_async` is a brilliant name.

It describes both of the main draws being the live state synchronization objects _and_ the `async` capability for (almost) all operations working better than the default IBKR `twsapi` multi-language API downloads.

i'll see what we can do with renaming some things and a github org (maybe the org name itself is one of the more descriptive names like `ib-api-reloaded`).

Since we know people _do_ have access to his original repos we could probably do a better cleanup job of shutting down/obsoleting the original package/repo/docs, but I personally don't want to bother them with anything else since they are going through enough and it's not their problem anymore.

**Reply by kulki007** [reply]
I agree lets go with ib_async.

**Reply by jlixfeld** [reply]
How do we feel about `insync_edw` as the GitHub organization name?

To paraphrase the original contributors from the `TWS API` list whose recent messages spurred this name:

- `insync` being included might make it more likely to be found by users who haven’t found this discussion.  While it's not in the library name, it's in the organization name so mission accomplished there.
- keeps the original name, recognises the legacy.

**Reply by mattsta** [reply]
Yeah, it'll be good to keep those references prominent probably more directly in docs and README so they remain more visible (and also builds up more search relevance over time).

I imagine if people searching for a library find the original archived `ib_insync` repo, they will search for `ib_insync replacement` somewhere too, so we'll make sure to include those details everywhere so they get picked up easily.

also it would be nice if we could find a copy of the original project announcement forum post from 2017? We'll add a new history/legacy document about how it started and how it's going.

**Reply by jlixfeld** [reply]
There was somebody who has the entire archive of the original list.  We can track them down to provide that information.

**Reply by vasiche** [reply]
If anybody has all or most/some legacy emails from groups.io I would like to collaborate with that person to see if we can somehow restore or make available to the community those emails.  Maybe saving them to a searchable pdf file, as a simple option.  There may be a better way, too.  Those emails had a lot of helpful information, answers, etc.  Let me know if you know who may have those emails.

**Comment by kulki007** [comment]
@mattsta How exactly does one create a PR if one doesnt have write permissions to the repository? Can anyone create a PR and then submit for approval and then your would approve the PR?

**Comment by kulki007** [comment]
@mattsta We need to agree on a new name for this project. We intend to create a new subgroup on groups.io but the name of the group should match the new name of this project. Do you any suggestions?

**Comment by mattsta** [comment]
they temporarily un-archived the repo, wrote [an official sad notice](https://github.com/erdewit/ib_insync/commit/ef5ea29e44e0c40bbadbc16c2281b3ac58aa4a40), then re-archived it again:

> Unfortunately, our dear brother Ewald de Wit passed away monday, March eleven as a result of a serious illness last year.
>
> Although we were not involved in his activities on this platform, we felt the due to inform you. We thank you for knowing him.

looks like we're on our own going forward. at least we can respect the legacy of his years of work by continuing to improve things over time.

**Comment by kulki007** [comment]
thank you for providing this update. This greatly makes me feel at ease. I wasnt willing to accept that Ewald somehow just went rogue. this makes more sense to me given what I know of him. RIP brother.
 I say we just go with a new project. Maybe we acknowledge the contribution that Ewald made by having some name like ib_insync_new or ib_insync_continued or soemthing like that.

**Comment by mattsta** [comment]
Status for now:

- made new org at https://github.com/ib-api-reloaded — let me know if you want to be added as a helper/contributor and we'll figure out org co-ownership eventually too, but i'll get the first re-branding and renaming steps done before making the new repo public. (again, priority is: repo/code stability, then public packaging, then public docs website)
- It will take a couple days to get the current code renamed and docs updated again. Maybe I'll have more time over the weekend. We'll also need to hope there are never any bugs in his other personal dependencies of `eventkit` and `nest_asyncio`, but I'm less concerned with `nest_asyncio` because it's mostly used for just ipython demos and I think [newer jupyter may not even need it anymore](https://stackoverflow.com/questions/47518874/how-do-i-run-python-asyncio-code-in-a-jupyter-notebook).

**Reply by jlixfeld** [reply]
I think there was a misunderstanding - the consensus was that the org name was to be `insync_edw`.  There seems to be a bunch of support for that on Mel's Discord.  I didn't get a chance to communicate that consensus here, apologies.

Also, I'll happily help out with co-ownership if there's any value. I'm not the most seasoned, but I'm happy to learn.

**Reply by ReliefCrew** [reply]
> the consensus was that the org name was to be ...

I didn't notice myself, but I still applaud the initiative. @mattsta... why not send a few invites to the top prospects?

**Reply by kulki007** [reply]
@mattsta can you set the org name to insync_edw as a tribute to Ewald who worked tirelessly all these years for free and helped so many of us. The actual project can be ib_async like you suggested.

**Reply by geepeem** [reply]
Great to see some progress! Thanks for spearheading this effort @mattsta . I am happy to help however I can, but am more of a trader than a coder TBH. Let me know if there's anything not too technical that needs doing. 
It would be great if we could get a good collaboration going with the IB dev team - it is definitely in their interest to make this project work since it makes using their API a lot simpler for the (probably many) people like me that just sort of picked up python as needed for creating their bots.

**Comment by cheukhin1024** [comment]
@mattsta Please let me know what I can do. I am working in IB API team.

**Reply by kulki007** [reply]
one of the issues I face is that we dont have access to a git repository for IB API itself. It will make matters so much easier if we can track changes on your end and then we replicate those changes on our end. Without a source control system, we need to manually diff files to figure out what exactly has changed between versions. Currently the source code is released in a windows installed file format which makes it very clunky to use. I hope you would agree its a lot simpler if it were just on github

**Reply by cheukhin1024** [reply]
@kulki007 Thank you for your reply. Unfortunately, I cannot make any promises here because I need to follow Interactive Brokers Communication Policy Guidelines and I am not the decision maker.

However, I sent an email to Mr. Matt (matt@matt.sh) using cXXXn@interactivebrokers.com and CC API team ( api@interactivebrokers.com) about introducing our new IBKR campus API documentation sites and showing my respect for Mr. Ewald de Wit and his contributions to the ib_insync project.

In my personal opinion, any possible ideas should be sent to Mr. Matt if he is confirmed to be the new main creator and the main contact person for ib_insync 2.0. After that, Mr. Matt can summarize the ideas and email to api@interactivebrokers.com .

**Reply by kulki007** [reply]
@mattsta I think the general consensus seems to be that we pay our respects to Ewald by creating the new github organization as insync_edw.
We should call the new project as ib_async as you had suggested. If you agree then we can create a Groups.IO account as well under the sub group called ib_async to match the  project name.

**Comment by mattsta** [comment]
Sunday update: over the weekend I renamed everything in the repo from `ib_insync` to `ib_async` and modernized some of the repo architecture.

New resources:

- new primary repo: https://github.com/ib-api-reloaded/ib_async (no new package release yet though)
- new docs site: https://ib-api-reloaded.github.io/ib_async/
- place for memories if people would like to add more details: https://github.com/ib-api-reloaded/memorial

Feel free to point out typos or anything else I missed in the refactored repo.

The new org name seems fine since we are making it clear we are continuing the project with all the original details. we are a library of peace. It seems cleaner with an unrelated org name unrelated to the previous activity because even at first glance, it may look like things named `edw` could still be live activity? please don't name anything after me eventually either.

**Reply by gnzsnz** [reply]
thanks for your effort on this.

I'm happy to help on keeping this running.

any plan to have a mail list in groups.io for example?

**Reply by geepeem** [reply]
This is great work Matt, thanks on behalf of all the people who used ib_insync and hopefully will be able to continue doing so thanks to your efforts!
The new repo looks great, and I agree that it's best to not include edw in the name to avoid any potential confusion. The memorial page is a nice touch.
I will read through the new repo and let you know if I spot anything that's worth changing / correcting.
Going forward, we should probably switch the discussion to the new repo. I will open one to get the ball rolling.
Cheers!

**Reply by ClimberMel** [reply]
I'm happy to help out with code.  Has anyone started on the changes for the fund data?  If not, you could assign that one to me.  Or if you add me to the repo, I can just assign it to myself and start on it.

**Reply by vasiche** [reply]
Thank you @mattsta!

**Reply by kulki007** [reply]
thank you so much @mattsta

**Reply by brndnmtthws** [reply]
As a suggestion, it might be worthwhile linking to https://github.com/ib-api-reloaded/ib_async at the top of the README and in the repo description/link.

**Comment by iamshanvy** [comment]
Hey @mattsta 

Thank you for creating a fork of repo so quickly, we are building custom solutions for our clients using `ib_insync` for a while now with some of them having their customised version of the library deployed, feel free to reach us for any help too, we would love to contribute the repo we have necessary resources to do that

Shameless plug, if anyone needs a custom bot developed using ib_insync (or `ib_async` now) reach us out [here](https://www.krunchdata.io/)

**Comment by decoded-markets** [comment]
First of all, RIP Edwin. Just came across this thread as I was looking for some info on the  ib_insync group and wondered why it was closed... Secondly, thanks heaps @mattsta for the initiative to continue the project. Like many, ib_insync -> ib_async  holds a critical place in my automated trading system. I would be more than happy to help contribute. I'm not sure my knowledge of how the library works behind the scene is up to scratch for now... but if push come to shove I'll spend the time it takes to make sure this library keeps on working. There is NO WAY I am starting an integration from scratch given my current IB setup has been working SO PERFECTLY thus far.

**Comment by westonplatter** [comment]
Looking forward to continuing to use ib_insync from the new ib-api-reloaded source.
